File: src/Dotflik.WebApp.Client/Interop/modalBootstrap.ts
/// <reference types="jquery" />
/// <reference types="bootstrap" />

/**
  * Hide a modal with id `modalId` asynchronously.
  *
  * @remarks
  * This function will wait for the modal to be hidden before it ends.
  *
  * @param modalId - The id of a modal dialog
  */
export async function hideModal(modalId: string): Promise<void> {
  let modalHidden: boolean = false;
  const eventName: string = "hidden.bs.modal";

  const hiddenHandler = function () {
    modalHidden = true;
  };

  $(modalId).on(eventName, hiddenHandler);

  $(modalId).modal("hide");
  while (!modalHidden) {
    await sleep(100);
  }

  $(modalId).off(eventName, hiddenHandler);
}

/**
  * Show a modal with id `modalId` asynchronously.
  * 
  * @remarks
  * This function will wait for the modal to be shown before it ends.
  *
  * @param modalId - The id of a modal dialog
  */
export async function showModal(modalId: string): Promise<void> {
  let modalShown: boolean = false;
  const eventName: string = "shown.bs.modal";

  const shownHandler = function () {
    modalShown = true;
  };

  $(modalId).on(eventName, shownHandler);

  $(modalId).modal("show");
  while (!modalShown) {
    await sleep(100);
  }

  $(modalId).off(eventName, shownHandler);
}

/**
  * Sleep for `ms` milliseconds.
  * @param ms - The number of milliseconds to sleep
  */
async function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

File: src/Dotflik.WebApp.Client/Interop/ModalBootstrap.cs
using System;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Dotflik.WebApp.Client.Interop
{
  /// <summary>
  /// This class exposes the functions defined in the module "Interop/modalBootstrap.ts".
  /// See https://getbootstrap.com/docs/4.0/components/modal/#usage.
  /// All methods of this class are asynchronous.
  /// </summary>
  public class ModalBootstrap : JavascriptModule
  {
    /// <inheritdoc/>
    protected override string ModuleScriptPath => "./js/modalBootstrap.js";

    /// <inheritdoc/>
    public ModalBootstrap(IJSRuntime jSRuntime) : base(jSRuntime) { }

    /// <summary>
    /// Show a modal with id <paramref name="modalId"/>.
    /// </summary>
    /// <remarks>
    /// This method will wait for the modal to be shown before it ends.
    /// </remarks>
    /// <param name="modalId">Id of the modal</param>
    /// <returns>Empty task</returns>
    public async Task ShowModalAsync(string modalId)
    {
      var normalizedModalId = NormalizeModalId(modalId);
      await Module.InvokeVoidAsync("showModal", normalizedModalId); 
    }

    /// <summary>
    /// Hide a modal with id <paramref name="modalId"/>.
    /// </summary>
    /// <remarks>
    /// This method will wait for the modal to be hidden before it ends.
    /// </remarks>
    /// <param name="modalId">Id of the modal</param>
    /// <returns>Empty task</returns>
    public async Task HideModalAsync(string modalId)
    {
      var normalizedModalId = NormalizeModalId(modalId);
      await Module.InvokeVoidAsync("hideModal", normalizedModalId);
    }

    /// <summary>
    /// Normalize the modal <paramref name="id"/> by prepending "#" to
    /// <paramref name="id"/>. If "#" is already prepended, then 
    /// <paramref name="id"/> is returned.
    /// </summary>
    /// <param name="id">The modal id</param>
    /// <returns>Return <paramref name="id"/> with "#" prepended to it</returns>
    protected static string NormalizeModalId(string id)
      => id.StartsWith("#") ? id : $"#{id}";

  }
}
